### ISSUE TITLE
needing -- before add-on command options is non-obvious

### ISSUE THREAD

**Issue by peti** [simonmichael/hledger#458]
[`hledger ui --watch` does not work] but `hledger-ui --watch` does. Apparently, `hledger` doesn't know about that flag and thus refuses to start up.

**Comment by simonmichael** [comment]
There are some [tricks required with options](http://hledger.org/hledger.html#options) in some cases. Do you need to insert a `--` ?

**Comment by peti** [comment]
I see. I didn't know about the need to terminate `hledger`'s option parsing explicitly. I understand why that's necessary from the PoV of someone who implements these kinds of things, too, but from a user's PoV this is rather unintuitive. Why do I need that when I call `hledger ui` but don't when I call `hledger bal`? The reason is that the `ui` command is implemented differently, but -- as a user -- I wouldn't really want to know (or think) about that detail.

**Comment by simonmichael** [comment]
I heartily agree! It's annoying. Any help welcome. The problem, which hledger tries and fails to completely hide, is that cmdargs (last I checked) isn't good at partially parsing the command line (passing though unrecognised options without complaining).

**Comment by peti** [comment]
As far as I understand it, there are 2 cases hledger has to deal with:

1) The user ran an internal command. In that case, hledger must parse and verify all options.

2) The user ran an external command. In that case, hledger must parse *none* of the command-line -- everything is just passed through to the backend via `exec()`.

Is that right?

**Comment by simonmichael** [comment]
That's what we'd like to happen, but AFAIK cmdargs doesn't support it yet. It insists on recognising everything in the command line, up to `--`. When I looked, I found it hard to see how to fix that, so instead I've tried workarounds like munging the command line before running cmdargs. But these imperfect, because doing that properly requires full knowledge of the options. But some time has passed; maybe cmdargs has changed ? My issue report should be in its tracker.

**Comment by peti** [comment]
How about dropping all argv elements that begin with a `-` character, taking the `head` of the remaining list, and checking that string (which should be the "command" to be run) against a list of known internal commands? If the string is not in that list, then `cmdargs` never needs to run at all.

**Comment by simonmichael** [comment]
We do something like that but it breaks down in case of `-F VALUE` or `--FLAG VALUE`.

**Comment by peti** [comment]
Yeah, that's true. And these kind of options can't be handled unless you know which ones take an argument either. So I guess using a separate parsing process is not feasible. 😞

**Comment by Xitian9** [comment]
I'm having a bit of difficulty determining what the issue is here. Is this issue obsolete or is there still something to fix?

**Comment by simonmichael** [comment]
Good news! The fix for #2388 was the final piece to remove this requirement. The `--` argument before addon options is no longer needed. \o/
